File: backend/app/schemas/bookmarks.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.core.reminders import normalize_reminder_offsets
from app.models.enums import BookmarkTarget, SeriesStatus
from app.schemas.schedule import DateTimeWithTimezone


class BookmarkCreate(BaseModel):
    target_type: BookmarkTarget
    target_id: UUID
    reminder_offsets: list[int] | None = None


class BookmarkUpdate(BaseModel):
    reminder_offsets: list[int]


class GuestBookmarkItem(BaseModel):
    target_type: BookmarkTarget
    target_id: UUID
    reminder_offsets: list[int] | None = None


class BookmarkMigrateBody(BaseModel):
    items: list[GuestBookmarkItem] = Field(default_factory=list, max_length=100)


class BookmarkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    target_type: BookmarkTarget
    target_id: UUID
    reminder_offsets: list[int]
    created_at: datetime


class BookmarkMigrateResponse(BaseModel):
    created: int
    skipped: int
    items: list[BookmarkRead]


class BookmarkDisplayFields(BaseModel):
    """Structured display fields shared by overview and guest resolve."""

    series_id: UUID | None
    event_id: UUID | None
    series_name: str
    series_status: SeriesStatus
    series_starts_on: date
    series_ends_on: date
    organizer_name: str
    organizer_slug: str
    venue_name: str
    venue_city: str
    event_number: int | None = None
    event_name: str | None = None
    flight_label: str | None = None
    nearest_start_at: DateTimeWithTimezone | None
    url: str


class BookmarkOverviewItem(BookmarkDisplayFields):
    id: UUID
    target_type: BookmarkTarget
    target_id: UUID
    reminder_offsets: list[int]
    created_at: datetime
    # Backward-compatible string fields for older clients.
    title: str
    subtitle: str | None


class BookmarkTargetResolveRef(BaseModel):
    target_type: BookmarkTarget
    target_id: UUID


class BookmarkTargetResolveBody(BaseModel):
    items: list[BookmarkTargetResolveRef] = Field(default_factory=list, max_length=100)


class BookmarkTargetResolveItem(BaseModel):
    target_type: BookmarkTarget
    target_id: UUID
    found: bool
    display: BookmarkDisplayFields | None = None


class BookmarkTargetResolveResponse(BaseModel):
    items: list[BookmarkTargetResolveItem]


def resolve_reminder_offsets(
    target_type: BookmarkTarget,
    offsets: list[int] | None,
    *,
    profile_defaults: list[int],
) -> list[int]:
    if target_type == BookmarkTarget.SERIES:
        if offsets is None:
            return []
        return normalize_reminder_offsets(offsets, allow_empty=True)
    if offsets is None:
        return normalize_reminder_offsets(profile_defaults, allow_empty=False)
    return normalize_reminder_offsets(offsets, allow_empty=False)

File: frontend/src/api/types/bookmarks.ts
import type { DateTimeWithTimezone, SeriesStatus } from "@/api/types/schedule";

export type BookmarkTargetType = "series" | "flight";

export interface Bookmark {
  id: string;
  target_type: BookmarkTargetType;
  target_id: string;
  reminder_offsets: number[];
  created_at: string;
}

export interface BookmarkDisplayFields {
  series_id: string | null;
  event_id: string | null;
  series_name: string;
  series_status: SeriesStatus;
  series_starts_on: string;
  series_ends_on: string;
  organizer_name: string;
  organizer_slug: string;
  venue_name: string;
  venue_city: string;
  event_number: number | null;
  event_name: string | null;
  flight_label: string | null;
  nearest_start_at: DateTimeWithTimezone | null;
  url: string;
}

export interface BookmarkOverviewItem extends BookmarkDisplayFields {
  id: string;
  target_type: BookmarkTargetType;
  target_id: string;
  reminder_offsets: number[];
  created_at: string;
  title: string;
  subtitle: string | null;
}

export interface BookmarkTargetResolveRef {
  target_type: BookmarkTargetType;
  target_id: string;
}

export interface BookmarkTargetResolveItem {
  target_type: BookmarkTargetType;
  target_id: string;
  found: boolean;
  display: BookmarkDisplayFields | null;
}

export interface BookmarkTargetResolvePayload {
  items: BookmarkTargetResolveRef[];
}

export interface BookmarkTargetResolveResponse {
  items: BookmarkTargetResolveItem[];
}

export interface BookmarkCreatePayload {
  target_type: BookmarkTargetType;
  target_id: string;
  reminder_offsets?: number[];
}

export interface BookmarkUpdatePayload {
  reminder_offsets: number[];
}

export interface GuestBookmarkItem {
  target_type: BookmarkTargetType;
  target_id: string;
  reminder_offsets?: number[];
}

export interface BookmarkMigratePayload {
  items: GuestBookmarkItem[];
}

export interface BookmarkMigrateResponse {
  created: number;
  skipped: number;
  items: Bookmark[];
}

/** Normalized row for bookmarks UI (server overview or guest hydrate). */
export interface BookmarkListItem extends BookmarkDisplayFields {
  key: string;
  source: "server" | "guest" | "demo";
  bookmarkId: string | null;
  target_type: BookmarkTargetType;
  target_id: string;
  reminder_offsets: number[];
  created_at: string;
}
